core/src/main/scala-3/sttp/tapir/macros/SchemaMacros.scala

```diff
@@ -3,8 +3,71 @@ package sttp.tapir.macros
 import sttp.tapir.Schema
 import sttp.tapir.generic.Configuration
 
-trait SchemaMacros[T] {
-  def modify[U](path: T => U)(modification: Schema[U] => Schema[U]): Schema[T] = ??? // TODO
+import scala.quoted.*
+
+trait SchemaMacros[T] { this: Schema[T] =>
+  inline def modify[U](inline path: T => U)(inline modification: Schema[U] => Schema[U]): Schema[T] = ${ SchemaMacros.modifyImpl[T, U]('this)('path)('modification)}
+}
+
+object SchemaMacros {
+  private val ShapeInfo = "Path must have shape: _.field1.field2.each.field3.(...)"
+
+  def modifyImpl[T: Type, U: Type](base: Expr[Schema[T]])(path: Expr[T => U])(modification: Expr[Schema[U] => Schema[U]])(using Quotes): Expr[Schema[T]] = {
+    import quotes.reflect.*
+    
+    enum PathElement {
+      case TermPathElement(term: String, xargs: String*) extends PathElement
+      case FunctorPathElement(functor: String, method: String, xargs: String*) extends PathElement
+    }
+
+    def toPath(tree: Tree, acc: List[PathElement]): Seq[PathElement] = {
+      def typeSupported(quicklensType: String) =
+        Seq("ModifyEach", "ModifyEither", "ModifyEachMap")
+          .exists(quicklensType.endsWith)
+      
+      tree match {
+        /** Field access */
+        case Select(deep, ident) =>
+          toPath(deep, PathElement.TermPathElement(ident) :: acc)
+        /** Method call with no arguments and using clause */
+        case Apply(Apply(TypeApply(Ident(f), a), idents), b) if typeSupported(f) => {
+           val newAcc = acc match {
+            /** replace the term controlled by quicklens */
+            case PathElement.TermPathElement(term, xargs @ _*) :: rest => PathElement.FunctorPathElement(f, term, xargs: _*) :: rest
+            case pathEl :: _ =>
+              report.throwError(s"Invalid use of path element $pathEl. $ShapeInfo, got: ${tree}")
+            case Nil =>
+              report.throwError(s"Invalid use of path element(Nil). $ShapeInfo, got: ${tree}")
```

Comment:
what case is it? probably error  message could be better

Reply:
hm I just copied it from scala 2 version and haven't found test example for it yet. I will try to add some.

Reply:
ah :) well maybe we could fix both at the same time ;) or maybe this case is impossible